src/ledger.ts

```diff
@@ -1280,3 +1310,237 @@ export class LedgerSignMessage extends LedgerBitcoinInteraction {
     });
   }
 }
+
+interface RegistrationConstructorArguments {
+  name: string;
+  braid: Braid;
+  policyHmac?: string;
+}
+
+/**
+ * A base class for any interactions that need to interact with a registered wallet
+ * by providing a base constructor that will generate the key origins and the policy
+ * from a given braid as well as methods for registering and returning a policy hmac
+ */
+export class LedgerBitcoinV2WithRegistrationInteraction extends LedgerBitcoinInteraction {
+  walletPolicy: MultisigWalletPolicy;
+
+  policyHmac?: Buffer;
+
+  policyId?: Buffer;
+
+  isLegacySupported = false;
+
+  isV2Supported = true;
+
+  constructor({ name, braid, policyHmac }: RegistrationConstructorArguments) {
+    super();
+    const keyOrigins = getKeyOriginsFromBraid(braid);
+    const template = getPolicyTemplateFromBraid(braid);
+    if (policyHmac) {
+      const error = validateHex(policyHmac);
+      if (error) throw new Error(`Invalid policyHmac`);
+      // TODO validate length
+      this.policyHmac = Buffer.from(policyHmac, "hex");
+    }
+    this.walletPolicy = new MultisigWalletPolicy({
+      name,
+      keyOrigins,
+      template,
+    });
+  }
+
+  messages() {
+    const messages = super.messages();
+
+    if (!this.policyHmac) {
+      messages.push({
+        state: ACTIVE,
+        level: INFO,
+        code: "ledger.confirm.address",
+        version: ">=2.1.0",
+        text: `Your Ledger will ask you to register your wallet info first. This allows the device to derive an address for a verified quorum.`,
+        action: LEDGER_RIGHT_BUTTON,
+      });
+    }
+
+    return messages;
+  }
+
+  registerWallet(verify = false): Promise<Buffer> {
+    if (this.policyHmac && !verify) return Promise.resolve(this.policyHmac);
+
+    // if we don't have a registered policy yet, then let's handle that
+    return this.withApp(async (app: AppClient) => {
+      const policy = this.walletPolicy.toLedgerPolicy();
+      const [policyId, policyHmac] = await app.registerWallet(policy);
+      const buff = Buffer.from(policyHmac);
+
+      if (
+        verify &&
+        this.policyHmac &&
+        this.policyHmac.toString("hex") !== buff.toString("hex")
+      ) {
+        console.error(
+          `Policy registrations did not match. Expected ${this.policyHmac.toString(
+            "hex"
+          )}; Actual: ${buff.toString("hex")}`
+        );
+      }
+
+      this.policyHmac = buff;
+      this.policyId = policyId;
+
+      return buff;
+    });
+  }
+}
+
+interface LedgerRegisterWalletPolicyArguments
+  extends RegistrationConstructorArguments {
+  verify?: boolean;
+}
+export class LedgerRegisterWalletPolicy extends LedgerBitcoinV2WithRegistrationInteraction {
+  verify: boolean;
```

Comment:
Same as above. Also, this is a better example for why to use private attributes but with getters/setters. Should any consumer be allowed to change this at any time or when the class is in any state, or should there be controls? Even if it doesn't need business logic controls now, having a getter/setter while keeping this private will indicate to other developers where the source of truth is controlled. If no gettter/setter, should this be `readonly`?